Phoenixcoin v0.7

sats0k opened the pull request.

sats0k committed:
Replace 'QT::SystemLocaleShortDate' by QLocale

‘Qt::SystemLocaleShortDate’ is deprecated

sats0k committed:
Use uint32_t integer type

uint32_t is a numeric type that guarantees 32 bits.

sats0k committed:
Replace Get64() by GetLow64()

The function Get64(.) has a bug in case the width is not divisible by 64.
Since it is only ever used as Get64(0) this simply changes it to this
special case.

From commit 3d34cb8eb5a7f5290579bd733f9514d253a5b87f

sats0k committed:
Convert memset calls to OPENSSL_cleanse

to securely erase sensitive keys values from memory.

sats0k committed:
Fix bits in ipv6 address groups

sats0k committed:
Disable deprecated-declarations


Allows to compile with OpenSSL v3.0

sats0k committed:
SHA256 APIs directly to EVP APIs


Switch from the low-level SHA256 APIs to EVP APIs.
The SHA256_* fuctions are deprecated on OpenSSL v3.0

sats0k committed:
Add missing files

src/leveldb.cpp
src/leveldb.h
src/leveldb/Makefile

sats0k committed:
Fix BN_zero() error

error: could not convert ‘BN_zero_ex(zero)’ from ‘void’ to ‘bool’

And correction of MiniUPnP previous commit.

ghostlander commented on src/util.h:
```diff
@@ -379,32 +381,39 @@ inline uint256 Hash(const T1 pbegin, const T1 pend)
 {
 private:
-    SHA256_CTX ctx;
+    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
```
warning: non-static data member initializers only available with -std=c++11 or -std=gnu++11
     EVP_MD_CTX *ctx = EVP_MD_CTX_new();

sats0k replied:
EVP functions are slower than the original functions that's why i stopped and real problems start when you try to replace RIPEMD160 which is well know for his length attack and must be replaced immediately.

ghostlander replied:
Can we keep the old API functions for OpenSSL v1.x.x and implement the EVP ones for v3.x only? OpenSSL calls consume a large share of CPU time as they are now.

sats0k replied:
This commit will be reverted to the old API functions.

ghostlander commented on src/db.h:
```diff
@@ -136,23 +136,23 @@ class CDB
-            return false;
+        OPENSSL_cleanse(datKey.get_data(), datKey.get_size());
+        bool success = false;
+        if (datValue.get_data() != nullptr) {
```
warning: identifier ‘nullptr’ is a keyword in C++11 [-Wc++0x-compat]
         if (datValue.get_data() != nullptr) {

sats0k replied:
i think you must add to .pro file 
`QMAKE_CXXFLAGS += -std=c++0x
`

ghostlander replied:
I know that. I use GCC as old as v4.2.1 for compatibility reasons. That's why I keep the whole code C++98/03 compatible.

sats0k replied:
For old version below GCC v4.7 replace nullptr with 0

ghostlander replied:
Why not keep NULL as it is now? Current GCC doesn't seem to complain.

sats0k replied:
You can use NULL, or 0, if you want, but nullptr is there for several good reasons
or if you want you can use something like

`#if __cplusplus < 201103L`
`#define nullptr 0`
`#endif `

sats0k committed:
Revert "SHA256 APIs directly to EVP APIs"

This reverts commit 3e6f99f14705d93e180edb425b1a636eaa68e26c.

sats0k committed:
Revert "Disable deprecated-declarations"

This reverts commit 12fd7fe3b740d3b48b1076869999c7a806d4079c.

sats0k committed:
Revert "Add missing files"

This reverts commit 8bb0afef5caeb093e05664feb22aa7afc55156f3.

sats0k committed:
Boost v1.87 compatible

Deprecated: No replacement.
is_v4_compatible()
is_v4_mapped()
So i disable the code.
